db/deletefile_test.cc

```diff
@@ -223,6 +223,54 @@ TEST_F(DeleteFileTest, PurgeObsoleteFilesTest) {
   CheckFileTypeCounts(dbname_, 0, 1, 1);
 }
 
+TEST_F(DeleteFileTest, WaitForCompactWithWaitForPurgeOptionTest) {
+  Options options = CurrentOptions();
+  SetOptions(&options);
+  Destroy(options);
+  options.create_if_missing = true;
+  Reopen(options);
+
+  std::string first("0"), last("999999");
+  CompactRangeOptions compact_options;
+  compact_options.change_level = true;
+  compact_options.target_level = 2;
+  Slice first_slice(first), last_slice(last);
+
+  CreateTwoLevels();
+  Iterator* itr = nullptr;
+  ReadOptions read_options;
+  read_options.background_purge_on_iterator_cleanup = true;
+  itr = db_->NewIterator(read_options);
+  ASSERT_OK(itr->status());
+  ASSERT_OK(db_->CompactRange(compact_options, &first_slice, &last_slice));
+  SyncPoint::GetInstance()->LoadDependency(
+      {{"DeleteFileTest::WaitForPurgeTest", "DBImpl::BGWorkPurge:start"}});
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  test::SleepingBackgroundTask sleeping_task_before;
+  env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask,
+                 &sleeping_task_before, Env::Priority::HIGH);
+  delete itr;
+
+  // not purged yet
+  CheckFileTypeCounts(dbname_, 0, 3, 1);
+  sleeping_task_before.WakeUp();
+
+  auto waiting_for_compaction_thread = port::Thread([this]() {
+    WaitForCompactOptions wait_for_compact_options;
+    wait_for_compact_options.wait_for_purge = true;
+    Status s = dbfull()->WaitForCompact(wait_for_compact_options);
+    ASSERT_OK(s);
+  });
+  // still not purged yet because of sync point
+  CheckFileTypeCounts(dbname_, 0, 3, 1);
+  TEST_SYNC_POINT("DeleteFileTest::WaitForPurgeTest");
+  waiting_for_compaction_thread.join();
```

Comment:
Sorry I meant to finish thinking about this before submitting the review but didn't. There's something slightly unconventional here, which is that the test could pass without the fix if waiting_for_compaction_thread were simply slow. Usually what I would do is have a sync point in the wait portion of WaitForCompact() that unblocks the purge (perhaps via a callback that does `sleeping_task_before.WakeUp()`)